k9mail/src/main/java/com/fsck/k9/ui/messageview/MessageViewFragment.java

```diff
@@ -741,26 +744,34 @@ public void onMessageDataLoadFinished(LocalMessage message) {
 
             displayHeaderForLoadingMessage(message);
             mMessageView.setToLoadingState();
+            showProgressThreshold = 0L;
         }
 
         @Override
         public void onMessageDataLoadFailed() {
             Toast.makeText(getActivity(), R.string.status_loading_error, Toast.LENGTH_LONG).show();
+            showProgressThreshold = 0L;
         }
 
         @Override
         public void onMessageViewInfoLoadFinished(MessageViewInfo messageViewInfo) {
             showMessage(messageViewInfo);
+            showProgressThreshold = 0L;
         }
 
         @Override
         public void onMessageViewInfoLoadFailed(MessageViewInfo messageViewInfo) {
             showMessage(messageViewInfo);
+            showProgressThreshold = 0L;
         }
 
         @Override
         public void setLoadingProgress(int current, int max) {
-            mMessageView.setLoadingProgress(current, max);
+            if (showProgressThreshold == 0L) {
+                showProgressThreshold = System.nanoTime() + PROGRESS_THRESHOLD_NANOS;
+            } else if (System.nanoTime() > showProgressThreshold) {
```

Comment:
This will only show the progress UI if there is actual progress to report. If the time to the first progress callback is e.g. 2 seconds the progress indicator won't be shown before that.

Reply:
Nit: Once the threshold is crossed there's no need to get the elapsed time on subsequent callback invocations.

Reply:
Correct, this relies on actual progress being returned. The expectation is that progress is being reported based on processed filesize, and we report the filesize for this express purpose (see [https://github.com/k9mail/k-9/blob/master/k9mail/src/main/java/com/fsck/k9/ui/crypto/MessageCryptoHelper.java#L395](MessageCryptoHelper:395)).

Reply:
I changed the threshold to a Long and use null now while there is none, and 0L as trivial threshold that always passes the smaller-than check without getting the elapsed time.